libs/single-spa-angular/elements/src/types.ts

```diff
@@ -2,8 +2,40 @@ import { NgModuleRef } from '@angular/core';
 import { NgElement } from '@angular/elements';
 import { BaseSingleSpaAngularOptions } from 'single-spa-angular/internals';
 
-export interface BootstrappedSingleSpaAngularElementsOptions extends BaseSingleSpaAngularOptions {
+// implementation from https://spin.atomicobject.com/2018/09/10/javascript-concurrency/
+export class Mutex {
```

Comment:
We should not need a mutex for allowing multiple instances of the same parcel config to be created and used. In single-spa-react, single-spa-vue, etc we solve the problem just by ensuring that the `bootstrap`, `mount`, `update`, and `unmount` functions are given variables specific to the instance so that they don't interfere with each other. See https://github.com/single-spa/single-spa-react/blob/4d77396990aa68284d2dc82a23ff1b3d3a6d6ed6/src/single-spa-react.js#L139-L140 which shows how we track separate instances on the `opts` object without resorting to using mutexes that prevent concurrency.

Reply:
In fact we could not create different instances for the angular elements, because the web components are essentially singletons. In fact this use case was checked for in the bootstrap method, but the check didn't work - it was a bug. The process is asynchronous, so we have had a reproducible race condition, in which we had the module creation called multiple times - just adding the same parcel with the angular element multiple times in a component.

Reply:
> because the web components are essentially singletons

I don't understand this, could you clarify? Are you saying that Angular's implementation of the custom element does not allow for multiple instances of the custom element? If so, that is a bug/limitation of Angular and seems unrelated to single-spa-angular.

Reply:
It is not angular's fault, the browser does not allow for multiple instances - angular just defines a custom component through the browser API (customElement.define)[https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define] and this API throws exception if the element is already defined [NotSupportedException](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define#exceptions) The way it is done at least in the examples is to call the define-method in the enclosing angular module.

Something along the lines:
```
ngDoBootstrap(): void {
    console.log("doing bootstrap in module v1")
    if (!customElements.get('mandant-widget-v1')) {
      customElements.define(
        // This tag we've have provided in `options.template` when called `singleSpaAngularElements`.
        'mandant-widget-v1',
        createCustomElement(MandantWidgetSelectionComponentV1, {injector: this.injector})
      );
      console.log('customElement registered');
    } else {
      console.log('customElement alredy called');
    }
  }
```

So even if we create the module once again in the bootstrap function the component will not and should not be defined again, as the browser will not allow it. The angular part just generates a bunch of javascript, which is mapped to the new tag through the browser API. 

The current single-spa-angular elements implementation also checks if the module is set and does not bootstrap it again, or at least that's the intention of the code - alas as I already said, the implementation has a race condition problem, if we try to insert multiple components at once - the module will be overwritten sometimes or the bootstrap function will not be called in the other cases.

Reply:
Calling `customElements.define()` multiple times for the same custom element is not valid, but having multiple instances of the custom element in the DOM is valid. Your earlier comment had me thinking that Angular Elements allowed neither, but your most recent comment clarifies that that's not the case.

Coming back to the original comment here, I still don't see any reason why a mutex would be needed.